mne/encoding/feature.py

```diff
@@ -0,0 +1,250 @@
+"""A collection of classes for transforming, preprocessing, and fitting."""
+# Authors: Chris Holdgraf <choldgraf@gmail.com>
+#          Jona Sassenhagen <jona.sassenhagen@gmail.de>
+#          Jean-Remi King <jeanremi.king@gmail.com>
+#
+# License: BSD (3-clause)
+import numpy as np
+from scipy import sparse
+from ..utils import warn
+
+
+class FeatureDelayer(object):
+    """Transformer for creating delayed versions of an input stimulus.
+
+    Parameters
+    ----------
+    time_window : array, shape (2,)
+        Specifies a minimum/maximum delay in seconds. In this case, all
+        delays between `min(time_window)` and `max(time_window)`
+        will be calculated using `sfreq`.
+    delays: array of floats, shape (n_delays,)
+        The time (in seconds) of each delay for specifying
+        pre-defined delays. Negative means timepoints in the past,
+        positive means timepoints in the future.
```

Comment:
ah I see what you mean - I agree with you that this isn't an ideal API. We
could use a single `delays` parameter, that could either accept a window or
a list of delays. The question is how to tell if a list of 2 delays are 2
explicit delays, or a tmin/tmax parameter. I was trying to add in window
support since jona made a good case that it'd be important. But perhaps
that's for the future rERP PR?

On Tue, Aug 2, 2016 at 12:33 PM, Jean-Rémi KING <notifications@github.com>
wrote:

> In mne/encoding/feature.py
> <https://github.com/mne-tools/mne-python/pull/3310#discussion_r73221730>:
>
> > +from ..utils import warn
> > +
> > +
> > +class FeatureDelayer(object):
> > +    """Transformer for creating delayed versions of an input stimulus.
> > +
> > +    Parameters
> > +    ----------
> > +    time_window : array, shape (2,)
> > +        Specifies a minimum/maximum delay in seconds. In this case, all
> > +        delays between `min(time_window)` and `max(time_window)`
> > +        will be calculated using `sfreq`.
> > +    delays: array of floats, shape (n_delays,)
> > +        The time (in seconds) of each delay for specifying
> > +        pre-defined delays. Negative means timepoints in the past,
> > +        positive means timepoints in the future.
>
> Perhaps I don't understand something, but I have the impression these
> delays and time_window are redundant and create an API issue. I.e. if you
> pass delays, you can't use time_window and vice versa, right? In principle,
> functions shouldn't have mutually exclusive parameters.
>
> If, for convenience purposes, one needs to compute the list of delays from
> a float param, the function should test for the type of input and transform
> the float into a list of delays.
>
> For example, in the time frequency module, we do a type detection for the
> decim parameter:
>
>    - if decim is int, we transform it into slice(None, None, decim),
>    - if it's a slice, we do nothing. Then we apply the decimation by
>    doing X[..., decim].
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3310/files/15a5bb641b7f7a9edf70f48f531590d992a366ab#r73221730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwSHfTXWd0PdglEWCSu65KenBi9ritSks5qb5t3gaJpZM4I28Ut>
> .
>